### PR TITLE
Transaction Handling, Cleanup, Minor fix

### PR DESCRIPTION
**Pull request by Jotschi**
The main change is that i added ```OrientGraphFactory.getTx()``` and the needed classes.

I'll add additional information to #16

### CODE REVIEW

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java [hunk 1]
```diff
@@ -53,7 +53,7 @@
 import com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage;
 
 @Graph.OptIn(Graph.OptIn.SUITE_STRUCTURE_STANDARD)
-public final class OrientGraph implements Graph {
```
Comment (velo) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: I was thinking about removing the final as well, but then I realized the other graph implementation are final...

Any particular reason to do so?

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientTransactionalGraph.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package org.apache.tinkerpop.gremlin.orientdb;
+
+import org.apache.commons.configuration.Configuration;
+import org.apache.tinkerpop.gremlin.structure.Transaction;
+
+public class OrientTransactionalGraph extends OrientGraph {
```
Comment (velo) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientTransactionalGraph.java]: any reason to not do this inside OrientGraph?

Comment (Jotschi) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientTransactionalGraph.java]: Mh, nope. But you are right. Just implementing tx in OrientGraph also works just fine. I'll rebase the commit and update the PR after some testing.

Comment (velo) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientTransactionalGraph.java]: =)

**Comment by velo** [comment]
Hey, update OrientGraphFeatures to include transaction...

That will kick more tests =D

**Comment by velo** [comment]
can you merge with new master? now we do have travis CI enabled

**Comment by velo** [comment]
BTW, do you have integration tests numbers?
````
Tests run: 776, Failures: 33, Errors: 178, Skipped: 261
````

**Comment by mpollmeier** [comment]
thanks @Jotschi - looking great
as @velo said, can you please merge master in (there's a conflict in the pom.xml) and we'll get this merged.

**Comment by Jotschi** [comment]
@mpollmeier I have just rebased my commits

**Comment by velo** [comment]
The build failed...
````
testGraphTransactionOnNoTrxOrientGraph(org.apache.tinkerpop.gremlin.orientdb.OrientGraphTest)  Time elapsed: 0.011 sec  <<< ERROR!
java.lang.Exception: Unexpected exception, expected<java.lang.RuntimeException> but was<java.lang.AssertionError>
````

**Comment by Jotschi** [comment]
@velo I have now added another flag which is used to indicate that the created graph is a noTx graph. In those cases .tx() usage is pretty much very useless since all operations on the graph are atomic. The current BP implementation of orientdb also supports commit and rolllback on noTx graph instances but those calls are basically ignored. I implemented this in a similar fashion.

**Comment by mpollmeier** [comment]
Do you guys have an idea how to fix the test failures? 

When I run tests locally using `mvn install` it stalls after some time. The last output I get is this. I increased maven's heap space, no difference. 

```
Running org.apache.tinkerpop.gremlin.orientdb.gremlintest.structure.OrientGraphStructureStandardIT
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message transform method call failed at JPLISAgent.c line: 844
```

Note that `mvn test` doesn't run all tests.

**Comment by velo** [comment]
@mpollmeier integration tests are "safe" to ignore for now... since we agreed that until they are 100% ok is ok to be x% brokne =D

**Comment by mpollmeier** [comment]
in that case I have no idea what travis is complaining about. if I run `mvn test` i get no failures

**Comment by Jotschi** [comment]
@velo @mpollmeier Tests are now green.

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java [hunk 3]
```diff
@@ -107,7 +108,7 @@ public boolean supportsComputer() {
 
         @Override
         public boolean supportsTransactions() {
-            return false;
+            return transactionalGraph;
```
Comment (velo) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: hrmmmm, ODBFeatures should define the overall feature set this driver does or does not support.

so, IMHO, I think it should be hardcoded to true

Comment (Jotschi) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: I adopted the OrientGraphNoTx implementation:
https://github.com/orientechnologies/orientdb/blob/master/graphdb/src/main/java/com/tinkerpop/blueprints/impls/orient/OrientGraphNoTx.java#L102

Comment (mpollmeier) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: I agree with @velo - the tinkerpop3 doc suggests that you can ask the graph up front if it supports a feature or not, independent if we enabled transaction on this instance or not:
`graph.features().graph().supportsTransactions()`

http://tinkerpop.incubator.apache.org/docs/3.0.2-incubating/

As I don't have push access to your branch I'll merge this to master and change it there. Let me know if I'm missing anything.

Comment (mpollmeier) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: done in 9b3d47e

Comment (Jotschi) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: If we decide to implement it that way we should get rid of getNoTx() or add a configuration flag which is used to determine the transactional graph behaviour.

Comment (mpollmeier) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: I reverted my change due to test failures. Happy to keep the discussion alive. What's your thoughts @velo ?

Comment (velo) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/ODBFeatures.java]: I would need to learn a little bit more, but, I think features should reflect what we do or do not support.

That affect how the test suites from gremlin gremlin behave.

But, one thing last days thought me, ain't that simple to get gremlin suite running orientdb

**Pull request merged**